modules/imgproc/include/opencv2/imgproc.hpp

```diff
@@ -3980,11 +3980,18 @@ all. It returns the rotated rectangle in which the ellipse is inscribed. The fir
 is used. Developer should keep in mind that it is possible that the returned
 ellipse/rotatedRect data contains negative indices, due to the data points being close to the
 border of the containing Mat element.
+If the flag direct is true a direct ellipse fit algorithm is used decribed by @cite Fitzgibbon99
 
 @param points Input 2D point set, stored in std::vector\<\> or Mat
  */
 CV_EXPORTS_W RotatedRect fitEllipse( InputArray points );
 
+/** @overload
+@param points Input 2D point set, stored in std::vector\<\> or Mat
+@param direct Flag to use a direct ellipse fit algorithm
+*/
+CV_EXPORTS_AS(fitEllipse2) RotatedRect fitEllipse( InputArray points, bool direct );
```

Comment:
The implementation of the WSCG98 paper I linked to in #6544 does not fail. (Though I used Eigen to compute eigenvectors. If there's a numerical problem in whatever OpenCV uses to compute eigenvalues, a new bug should be raised.) That paper is a successor to the Fitzgibbon's paper and a critique of it because it points out inherent flaws in the Fitzgibbon's method:

> Apart from its theoretical correctness, the original Fitzgibbon’s approach described in the previous section has several drawbacks. The matrix C (Eq. 9) is singular and the matrix S (Eq. 11) is also nearly singular (it is singular if all data points lie exactly on an ellipse). Regarding that, the computation of the eigenvalues of Eq. 10 is numerically unstable and can produce wrong results (as inﬁnite or complex numbers). It should be noted that using of an inverse as proposed in the original code (see line 5 in Fig. 1) does not solve this problem.
>
>Another problematic part of the algorithm is a localization of the optimal solution of the ﬁtting. In [Fitzg96a] authors proved that Eq. 10 has exactly one positive eigenvalue and they stated that the corresponding eigenvector is an optimal solution of Eq. 7. Unfortunately, this is not true. In an ideal case, when all data points lie exactly on an ellipse, the eigenvalue is zero. Moreover, regarding a numerical computation of eigenvalues the optimal eigenvalue can even be a small negative number. In such situations, the proposed localization (see line 6 in Fig. 1) can produce non-optimal or completely wrong solutions.

Do you have anything against implementing the WSCG98 paper or even a newer method if you happen to find it? (I personally chose WSCG98 for my application because I needed something that worked and that worked fast, i.e., it was a real-time application.)